lightning/src/offers/invoice_request.rs

```diff
@@ -0,0 +1,1115 @@
+			.amount_msats(2_000)
+			.quantity(2).unwrap()
+			.build().unwrap()
+			.sign(payer_sign).unwrap();
+
+		let mut buffer = Vec::new();
+		invoice_request.write(&mut buffer).unwrap();
+
+		if let Err(e) = InvoiceRequest::try_from(buffer) {
+			panic!("error parsing invoice_request: {:?}", e);
+		}
+
+		let invoice_request = OfferBuilder::new("foo".into(), recipient_pubkey())
+			.amount_msats(1000)
+			.supported_quantity(Quantity::Unbounded)
+			.build().unwrap()
+			.request_invoice(vec![1; 32], payer_pubkey())
+			.build_unchecked()
+			.sign(payer_sign).unwrap();
+
+		let mut buffer = Vec::new();
+		invoice_request.write(&mut buffer).unwrap();
+
+		match InvoiceRequest::try_from(buffer) {
+			Ok(_) => panic!("expected error"),
+			Err(e) => assert_eq!(e, ParseError::InvalidSemantics(SemanticError::MissingQuantity)),
+		}
+	}
+
+	#[test]
+	fn fails_parsing_invoice_request_without_metadata() {
+		let offer = OfferBuilder::new("foo".into(), recipient_pubkey())
+			.amount_msats(1000)
+			.build().unwrap();
+		let unsigned_invoice_request = offer.request_invoice(vec![1; 32], payer_pubkey())
+			.build().unwrap();
+		let mut tlv_stream = unsigned_invoice_request.invoice_request.as_tlv_stream();
+		tlv_stream.0.metadata = None;
+
+		let mut buffer = Vec::new();
+		tlv_stream.write(&mut buffer).unwrap();
+
+		match InvoiceRequest::try_from(buffer) {
+			Ok(_) => panic!("expected error"),
+			Err(e) => {
+				assert_eq!(e, ParseError::InvalidSemantics(SemanticError::MissingPayerMetadata));
+			},
+		}
+	}
+
+	#[test]
+	fn fails_parsing_invoice_request_without_payer_id() {
+		let offer = OfferBuilder::new("foo".into(), recipient_pubkey())
+			.amount_msats(1000)
+			.build().unwrap();
+		let unsigned_invoice_request = offer.request_invoice(vec![1; 32], payer_pubkey())
+			.build().unwrap();
+		let mut tlv_stream = unsigned_invoice_request.invoice_request.as_tlv_stream();
+		tlv_stream.2.payer_id = None;
+
+		let mut buffer = Vec::new();
+		tlv_stream.write(&mut buffer).unwrap();
+
+		match InvoiceRequest::try_from(buffer) {
+			Ok(_) => panic!("expected error"),
+			Err(e) => assert_eq!(e, ParseError::InvalidSemantics(SemanticError::MissingPayerId)),
+		}
+	}
+
+	#[test]
+	fn fails_parsing_invoice_request_without_node_id() {
+		let offer = OfferBuilder::new("foo".into(), recipient_pubkey())
+			.amount_msats(1000)
+			.build().unwrap();
+		let unsigned_invoice_request = offer.request_invoice(vec![1; 32], payer_pubkey())
+			.build().unwrap();
+		let mut tlv_stream = unsigned_invoice_request.invoice_request.as_tlv_stream();
+		tlv_stream.1.node_id = None;
+
+		let mut buffer = Vec::new();
+		tlv_stream.write(&mut buffer).unwrap();
+
+		match InvoiceRequest::try_from(buffer) {
+			Ok(_) => panic!("expected error"),
+			Err(e) => {
+				assert_eq!(e, ParseError::InvalidSemantics(SemanticError::MissingSigningPubkey));
+			},
+		}
+	}
+
+	#[test]
+	fn fails_parsing_invoice_request_with_invalid_signature() {
```

Comment:
Looks like we don't currently test that we can parse an invreq without a signature, though I assume that'll be done in follow-up anyway